File: contest/ICPC/2018/欧洲/NWERC西北欧/data/brexitnegotiations/submissions/nlog2n_solutions/bjarki_bs.java
// Ported to Java by Per
import java.util.*;
import java.io.*;

class bjarki_bs {
    public static void main(String args[]) {
        Kattio io = new Kattio(System.in, System.out);
        int n = io.getInt();
        ArrayList<Integer> succ[] = new ArrayList[n], prec[] = new ArrayList[n];
        int len[] = new int[n];
        int lo = 0,
            hi = 0,
            res = -1;
        for (int i = 0; i < n; ++i) {
            succ[i] = new ArrayList<Integer>();
            prec[i] = new ArrayList<Integer>();
        }
        for (int i = 0; i < n; ++i) {
            len[i] = io.getInt();
            int cnt = io.getInt();
            lo = Math.max(lo, len[i]);
            hi = Math.max(hi, len[i]+n);
            for (int j = 0; j < cnt; ++j) {
                int x = io.getInt() - 1;
                succ[i].add(x);
                prec[x].add(i);
            }
        }

        while (lo <= hi) {
            int mx = (lo+hi)/2;
            ArrayList<Integer> pos[] = new ArrayList[n];
            int left[] = new int[n];
            for (int i = 0; i < n; ++i)
                pos[i] = new ArrayList<Integer>();
            for (int i = 0; i < n; ++i) {
                left[i] = prec[i].size();
                pos[Math.min(n-1, mx - len[i])].add(i);
            }

            TreeSet<Integer> waiting = new TreeSet<Integer>(), ready = new TreeSet<Integer>();
            boolean ok = true;
            for (int i = n-1; i >= 0; i--) {
                for (int x: pos[i]) {
                    if (left[x] == 0) {
                        ready.add(x);
                    } else {
                        waiting.add(x);
                    }
                }
                if (ready.isEmpty()) {
                    ok = false;
                    break;
                }
                int cur = ready.pollFirst();
                for (int x: succ[cur]) {
                    if (--left[x] == 0 && waiting.remove(x)) {
                        ready.add(x);
                    }
                }
            }
            if (ok) {
                res = mx;
                hi = mx-1;
            } else {
                lo = mx+1;
            }
        }

        io.println(res);
        io.flush();

    }
}

/** Simple yet moderately fast I/O routines.
 *
 * Example usage:
 *
 * Kattio io = new Kattio(System.in, System.out);
 *
 * while (io.hasMoreTokens()) {
 *    int n = io.getInt();
 *    double d = io.getDouble();
 *    double ans = d*n;
 *
 *    io.println("Answer: " + ans);
 * }
 *
 * io.close();
 *
 *
 * Some notes:
 *
 * - When done, you should always do io.close() or io.flush() on the
 *   Kattio-instance, otherwise, you may lose output.
 *
 * - The getInt(), getDouble(), and getLong() methods will throw an
 *   exception if there is no more data in the input, so it is generally
 *   a good idea to use hasMoreTokens() to check for end-of-file.
 *
 * @author: Kattis
 */

class Kattio extends PrintWriter {
    public Kattio(InputStream i) {
        super(new BufferedOutputStream(System.out));
        r = new BufferedReader(new InputStreamReader(i));
    }
    public Kattio(InputStream i, OutputStream o) {
        super(new BufferedOutputStream(o));
        r = new BufferedReader(new InputStreamReader(i));
    }

    public boolean hasMoreTokens() {
        return peekToken() != null;
    }

    public int getInt() {
        return Integer.parseInt(nextToken());
    }

    public double getDouble() {
        return Double.parseDouble(nextToken());
    }

    public long getLong() {
        return Long.parseLong(nextToken());
    }

    public String getWord() {
        return nextToken();
    }



    private BufferedReader r;
    private String line;
    private StringTokenizer st;
    private String token;

    private String peekToken() {
        if (token == null)
            try {
                while (st == null || !st.hasMoreTokens()) {
                    line = r.readLine();
                    if (line == null) return null;
                    st = new StringTokenizer(line);
                }
                token = st.nextToken();
            } catch (IOException e) { }
        return token;
    }

    private String nextToken() {
        String ans = peekToken();
        token = null;
        return ans;
    }
}

File: contest/ICPC/2018/欧洲/NWERC西北欧/data/brexitnegotiations/submissions/nlog2n_solutions/bjarki_bs.cpp
#include <bits/stdc++.h>
using namespace std;
template <class T> int size(const T &x) { return x.size(); }
#define rep(i,a,b) for (auto i=(a); i<(b); ++i)
#define iter(it,c) for (auto it = (c).begin(); it != (c).end(); ++it)
typedef pair<int, int> ii;
typedef vector<int> vi;
typedef vector<ii> vii;
typedef vector<vi> vvi;
typedef long long ll;
const int INF = 2147483647;

int main() {
    int n;
    cin >> n;
    vector<vi> succ(n), prec(n);
    vi len(n);
    int lo = 0,
        hi = 0,
        res = -1;
    rep(i,0,n) {
        int cnt;
        cin >> len[i] >> cnt;
        lo = max(lo, len[i]);
        hi = max(hi, len[i]+n);
        rep(j,0,cnt) {
            int x;
            cin >> x;
            x--;
            succ[i].push_back(x);
            prec[x].push_back(i);
        }
    }

    while (lo <= hi) {
        int mx = (lo+hi)/2;
        vector<vi> pos(n);
        vi left(n);
        rep(i,0,n) {
            left[i] = size(prec[i]);
            pos[min(n-1, mx - len[i])].push_back(i);
        }
        set<int> waiting, ready;
        bool ok = true;
        for (int i = n-1; i >= 0; i--) {
            iter(it,pos[i]) {
                if (left[*it] == 0) {
                    ready.insert(*it);
                } else {
                    waiting.insert(*it);
                }
            }
            if (ready.empty()) {
                ok = false;
                break;
            }
            int cur = *ready.begin();
            ready.erase(cur);
            iter(it,succ[cur]) {
                if (--left[*it] == 0 && waiting.find(*it) != waiting.end()) {
                    waiting.erase(*it);
                    ready.insert(*it);
                }
            }
        }
        if (ok) {
            res = mx;
            hi = mx-1;
        } else {
            lo = mx+1;
        }
    }

    cout << res << endl;

    return 0;
}

